### PR TITLE
Add more typing to airflow.utils.helpers

### PR DESCRIPTION
**Pull request by mik-laj**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/utils/helpers.py [hunk 1]
```diff
@@ -31,10 +31,17 @@
 from airflow.exceptions import AirflowException
 from airflow.utils.module_loading import import_string
 
+if TYPE_CHECKING:
+    from airflow.models import TaskInstance
+
 KEY_REGEX = re.compile(r'^[\w.-]+$')
+CAMELCASE_TO_SNAKE_CASE_REGEX = re.compile('(?!^)([A-Z]+)')
```
Comment (uranusjr) [inline airflow/utils/helpers.py]: I’m surprised for all the opinions pylint have, it does not bother to enforce a regex pattern must always use an r-string.

File: airflow/utils/helpers.py [hunk 2]
```diff
@@ -181,7 +184,7 @@ def merge_dicts(dict1, dict2):
     return merged
 
 
-def partition(pred: Callable, iterable: Iterable):
+def partition(pred: Callable[[T], bool], iterable: Iterable[T]):
```
Comment (uranusjr) [inline airflow/utils/helpers.py]:
```suggestion
def partition(pred: Callable[[T], bool], iterable: Iterable[T]) -> Tuple[Iterable[T], Iterable[T]]:
```

(I think?)

File: airflow/utils/helpers.py [hunk 3]
```diff
@@ -137,7 +140,7 @@ def parse_template_string(template_string):
         return template_string, None
 
 
-def render_log_filename(ti, try_number, filename_template):
+def render_log_filename(ti: "TaskInstance", try_number, filename_template):
```
Comment (uranusjr) [inline airflow/utils/helpers.py]:
```suggestion
def render_log_filename(ti: "TaskInstance", try_number, filename_template) -> str:
```

File: airflow/utils/helpers.py (hunk 1 above)
Comment (mik-laj) [inline airflow/utils/helpers.py]:
```suggestion
CAMELCASE_TO_SNAKE_CASE_REGEX = re.compile(r'(?!^)([A-Z]+)')
```

File: airflow/utils/helpers.py [hunk 4]
```diff
@@ -21,7 +21,7 @@
 from datetime import datetime
 from functools import reduce
 from itertools import filterfalse, tee
-from typing import Any, Callable, Dict, Generator, Iterable, List, Optional, TypeVar
+from typing import TYPE_CHECKING, Any, Callable, Dict, Generator, Iterable, List, Optional, TypeVar
```
Comment (XD-DENG) [inline airflow/utils/helpers.py]:
```suggestion
from typing import TYPE_CHECKING, Any, Callable, Dict, Generator, Iterable, List, Optional, Tuple, TypeVar
```

This is needed to address the static check error caused by line 187

**Review by XD-DENG — CHANGES_REQUESTED** [review]

File: airflow/utils/helpers.py (hunk 4 above)
Comment (mik-laj) [reply airflow/utils/helpers.py]: Thanks.Added.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
I guess it needs rebase @mik-laj

**Comment by jhtimmins** [comment]
@XD-DENG Can you take another look at validate that your requested changes have been addressed?

**Review by XD-DENG — APPROVED** [review]

**Pull request merged**